File: HackerRank/Prepare/Algorithms/Implementation/Matrix Layer Rotation.js
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'matrixRotation' function below.
 *
 * The function accepts following parameters:
 *  1. 2D_INTEGER_ARRAY matrix
 *  2. INTEGER r
 */

function matrixRotation(matrix, r) {
    const numberOfRows = matrix.length;
    const numberOfColumns = matrix[0].length;
    const deep = Math.floor(Math.min(numberOfRows, numberOfColumns)/2);
    for(let rep = 0; rep<r; rep++){
        for(let innerDeep = 0; innerDeep <=deep-1; innerDeep++){
        let auxRow;
        for(let i = innerDeep; i<numberOfRows - innerDeep; i++){
        const row = matrix[i];
        if(i === innerDeep){
            let aux = row[row.length-1-innerDeep]
            for(let j = row.length -1 - innerDeep; j>innerDeep; j--){
                let aux2 = aux;
                aux = row[j-1]
                row[j-1] = aux2
            }
            auxRow = aux;
        }
        if(i === matrix.length - 1 -innerDeep){
            let aux = auxRow
            for(let j = innerDeep; j<row.length - 1 - innerDeep; j++){
                let aux2 = aux;
                aux = row[j+1]
                row[j+1] = aux2
            }
            auxRow = aux;
        }else{
            let aux2 = auxRow
            auxRow = matrix[i+1][innerDeep]
            matrix[i+1][innerDeep] = aux2
        }
    }
    for(let i= matrix.length - 1 - innerDeep; i>innerDeep; i--){
        let aux2 = auxRow
        const totalLength = matrix[i-1].length
        auxRow = matrix[i-1][totalLength-1-innerDeep]
        matrix[i-1][totalLength-1-innerDeep] = aux2
    }
        }
    }
    
    const joinedMatrix = matrix.map(row => row.join(' ')).join('\n')
    console.log(joinedMatrix)

}

function main() {
    const firstMultipleInput = readLine().replace(/\s+$/g, '').split(' ');

    const m = parseInt(firstMultipleInput[0], 10);

    const n = parseInt(firstMultipleInput[1], 10);

    const r = parseInt(firstMultipleInput[2], 10);

    let matrix = Array(m);

    for (let i = 0; i < m; i++) {
        matrix[i] = readLine().replace(/\s+$/g, '').split(' ').map(matrixTemp => parseInt(matrixTemp, 10));
    }

    matrixRotation(matrix, r);
}

File: HackerRank/Prepare/Algorithms/Implementation/Matrix Layer Rotation.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'matrixRotation' function below.
#
# The function accepts following parameters:
#  1. 2D_INTEGER_ARRAY matrix
#  2. INTEGER r
#

def get_layer_points(
    m: int,
    n: int,
    layer: int,
) -> list[tuple[int, int]]:
    points: list[tuple[int, int]] = []
    for row in range(layer, m - layer - 1):
        points.append((row, layer))
    for col in range(layer, n - layer - 1):
        points.append((m - layer - 1, col))
    for row in range(m - layer - 1, layer, -1):
        points.append((row, n - layer - 1))
    for col in range(n - layer - 1, layer, -1):
        points.append((layer, col))
    return points

def matrixRotation(matrix: list[list[int]], r: int):
    # Write your code here
    m = len(matrix)
    n = len(matrix[0])
    output = [line.copy() for line in matrix]
    for layer in range(min(m, n) // 2):
        dst = get_layer_points(m, n, layer)
        steps = r % len(dst)
        src = dst[-steps:] + dst[:-steps]
        for i in range(len(dst)):
            output[dst[i][0]][dst[i][1]] = matrix[src[i][0]][src[i][1]]
    for line in output:
        print(*line, sep=" ")
    

if __name__ == '__main__':
    first_multiple_input = input().rstrip().split()

    m = int(first_multiple_input[0])

    n = int(first_multiple_input[1])

    r = int(first_multiple_input[2])

    matrix = []

    for _ in range(m):
        matrix.append(list(map(int, input().rstrip().split())))

    matrixRotation(matrix, r)